### PR TITLE
Rosters

### PR DESCRIPTION
**Pull request by Anbranin**
In testing your code I reviewed it as well, so I think it's fine for now. I fixed some things I found. What do you think of the tests?

### CODE REVIEW

**Comment by dfaulken** [comment]
Will do, looks like there are some Rubocop issues.

File: app/controllers/rosters_controller.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+class RostersController < ApplicationController
+  before_action :find_roster, only: [:destroy, :edit, :update]
+
+  def create
+    roster_params = params.require(:roster).permit(:name)
+    roster = Roster.new roster_params
+    if roster.save
+      flash[:message] = 'roster has been created.'
```
Comment (dfaulken) [inline app/controllers/rosters_controller.rb]: Initial capitalization

File: app/controllers/rosters_controller.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      redirect_to rosters_path
+    else
+      flash[:errors] = roster.errors.full_messages
+      redirect_to :back
+    end
+  end
+
+  def destroy
+    @roster.destroy
+    flash[:message] = 'roster and any assignments have been deleted.'
```
Comment (dfaulken) [inline app/controllers/rosters_controller.rb]: Initial capitalization

File: app/controllers/rosters_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    @rosters = Roster.all
+  end
+
+  def new
+  end
+
+  def update
+    roster_params = params.require(:roster).permit!
+    if @roster.update roster_params
+      flash[:message] = 'roster has been updated.'
```
Comment (dfaulken) [inline app/controllers/rosters_controller.rb]: Initial capitalization

File: spec/controllers/assignments_controller_spec.rb [hunk 4]
```diff
@@ -59,26 +63,27 @@
...
     end
     it 'finds the correct assignment' do
       submit
       expect(assigns.fetch :assignment).to eql @assignment
     end
     it 'populates a users variable of all users' do
-      user_1 = create :user
-      user_2 = create :user
-      user_3 = create :user
+      user_1 = create :user, rosters: [@roster]
```
Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: Do you think it would be worth writing a spec_helper method like `create_user` which you pass a roster, just to cut down on the repetition between lines here?

File: spec/controllers/assignments_controller_spec.rb [hunk 5]
```diff
@@ -90,25 +95,30 @@
...
            end_date: @end_date,
            user_ids: @user_ids,
            starting_user_id: @starting_user_id
     end
     it 'calls Assignment#generate rotation with the given arguments' do
-      expect(Assignment).to receive(:generate_rotation)
-        .with @user_ids, Date.today, Date.tomorrow, @starting_user_id
+      # remove all other instances of roster so 'any instance' definitely
+      # refers to our @roster instance
+      Roster.where.not(id: @roster.id).delete_all
```
Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: OK but why are there any if we're not creating them here

File: spec/controllers/rosters_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    let :submit do
+      post :create, roster: { name: 'Operations' }
+    end
+    context 'without errors' do
+      it 'creates a roster' do
+        expect{ submit }.to change{ Roster.count }.from(1).to(2)
+      end
+      it 'redirects to the index' do
+        submit
+        expect(response).to redirect_to(
```
Comment (dfaulken) [inline spec/controllers/rosters_controller_spec.rb]: Why not a single line?

File: spec/controllers/rosters_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    end
+    let :submit do
+      delete :destroy, id: @roster.id
+    end
+    it 'deletes the correct roster' do
+      submit
+      expect(Roster.where(id: @roster.id)).to be_empty
+    end
+    it 'destroys the roster' do
+      expect_any_instance_of(Roster)
```
Comment (dfaulken) [inline spec/controllers/rosters_controller_spec.rb]: Why not a single line?

File: spec/controllers/rosters_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    end
+    it 'renders the correct template' do
+      submit
+      expect(response).to render_template :index
+    end
+  end
+
+  describe 'GET #new' do
+    it 'renders the new template' do
+      expect(get :new).to render_template :new
```
Comment (dfaulken) [inline spec/controllers/rosters_controller_spec.rb]: Weird parentheses

File: spec/controllers/rosters_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    end
+    context 'without errors' do
+      it 'updates the roster' do
+        submit
+        expect(@roster.reload.name).to eql 'unique'
+        expect(@roster.reload.fallback_user).to eql @user
+      end
+      it 'redirects to the index' do
+        submit
+        expect(response).to redirect_to(
```
Comment (dfaulken) [inline spec/controllers/rosters_controller_spec.rb]: Why not a single line?

File: spec/factories/users.rb [hunk 10]
```diff
@@ -5,5 +5,6 @@
     sequence(:spire) { |n| n.to_s.rjust(8, '0') + '@umass.edu' }
     sequence(:email) { |n| "user#{n}@umass.edu" }
     sequence(:phone) { |n| '+1' + n.to_s.rjust(10, '0') }
+    rosters { [FactoryGirl.create(:roster)] }
```
Comment (dfaulken) [inline spec/factories/users.rb]: Does one need to say `FactoryGirl`?

File: spec/factories/assignments.rb [hunk 11]
```diff
@@ -1,6 +1,7 @@
 FactoryGirl.define do
   factory :assignment do
-    user
+    roster
+    user { FactoryGirl.create(:user, rosters: [roster]) }
```
Comment (dfaulken) [inline spec/factories/assignments.rb]: Does one need to say `FactoryGirl`?

File: spec/factories/users.rb (hunk 10 above)
Comment (dfaulken) [inline spec/factories/users.rb]: Also, holy brackets, :bat: 👨

File: spec/controllers/assignments_controller_spec.rb (hunk 4 above)
Comment (Anbranin) [inline spec/controllers/assignments_controller_spec.rb]: No, because there can't be any users without rosters.

File: spec/controllers/assignments_controller_spec.rb (hunk 5 above)
Comment (Anbranin) [inline spec/controllers/assignments_controller_spec.rb]: when_current_user_is creates a user. Creating a user creates a roster, since there can never be any users unless there are rosters.

File: spec/controllers/assignments_controller_spec.rb (hunk 4 above)
Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: I don't understand your answer.

My question was whether it would be worth writing a method e.g.:

```ruby
def roster_user(roster)
  create :user, rosters: [roster]
end
```

So that we could cut down on the repetition here, by calling e.g.:

```
user_1 = roster_user(@roster)
```

Comment (Anbranin) [inline spec/controllers/assignments_controller_spec.rb]: Then we'd literally never be calling create :user, except to test that it doesn't work without having a roster. My point is there _are_ no users that aren't roster_users.

Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: And *my* point is that the current code is extremely repetitive.

File: spec/controllers/rosters_controller_spec.rb (hunk 8 above)
Comment (Anbranin) [inline spec/controllers/rosters_controller_spec.rb]: What do you mean by 'weird'

File: spec/factories/assignments.rb (hunk 11 above)
Comment (Anbranin) [inline spec/factories/assignments.rb]: hmm guess not.

File: spec/controllers/rosters_controller_spec.rb (hunk 8 above)
Comment (dfaulken) [inline spec/controllers/rosters_controller_spec.rb]: Eh, never mind, it's not a big deal and Rubocop didn't flag it. Disregard.

File: spec/controllers/assignments_controller_spec.rb (hunk 4 above)
Comment (Anbranin) [inline spec/controllers/assignments_controller_spec.rb]: And I'm saying there's no good way to fix that.

Comment (Anbranin) [inline spec/controllers/assignments_controller_spec.rb]: "These 6 lines are repetitive, therefore let us write more stuff every single time we create a user" doesn't seem like a good argument to me.

Comment (Anbranin) [inline spec/controllers/assignments_controller_spec.rb]: That is, there are only a few instances when I want a user to belong to a specific roster, so it's repetitive in that case, but very few.

Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: Okay, well, I felt like I was reading the exact same code a bunch but if that's not the case then whatever.

Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: FWIW

```bash
grep -r "rosters: \[@roster\]" . | wc -l
```

gives me 23.

Comment (fermion) [inline spec/controllers/assignments_controller_spec.rb]: IMHO there's something to be said for WET tests cases. There's something to be said for total isolation for each example, even if comes at the expense of repetitive LOC.

That being said, if a macro makes sense that that's a way to go here. I think the danger there is that said macro can (and almost always will) grow moving more and more setup logic out of the individual examples making it harder to understand (in the future) why tests fail.

File: spec/controllers/assignments_controller_spec.rb [hunk 12]
```diff
@@ -81,9 +81,9 @@
       expect(assigns.fetch :assignment).to eql @assignment
     end
     it 'populates a users variable of all users' do
-      user_1 = create :user, rosters: [@roster] 
-      user_2 = create :user, rosters: [@roster]
-      user_3 = create :user, rosters: [@roster]
+      user_1 = roster_user @roster
```
Comment (fermion) [inline spec/controllers/assignments_controller_spec.rb]: This is really no better or worse than before :trollface:

File: spec/controllers/assignments_controller_spec.rb (hunk 4 above)
Comment (dfaulken) [inline spec/controllers/assignments_controller_spec.rb]: In principle that's true. This example is pretty trivial isolation. If this 'macro' fails, we've got bigger problems.

File: .rubocop.yml [hunk 13]
```diff
@@ -1,2 +1,5 @@
 require: umts-custom-cops
 inherit_from: .rubocop-no-umts-cops.yml
+
+HasAndBelongsToMany:
```
Comment (dfaulken) [inline .rubocop.yml]: Personally, I like to add comments explaining why we're disagreeing with Rubocop.
